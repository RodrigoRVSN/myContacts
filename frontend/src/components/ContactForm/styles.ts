import styled from 'styled-components';

export const Form = styled.form`
  margin-bottom: 1rem;
  color: ${({ theme }) => theme.colors?.primary.lighter};

    a {
      display: flex;
      align-items: center;
      text-decoration: none;

      span{
        color: ${({ theme }) => theme.colors?.primary.light};
        font-weight: bold;
      }

      img{
        margin-right: 0.5rem;
        transform: rotate(-90deg);
      }
    }
`;

export const ButtonContainer = styled.div`
  margin-top: 2rem;

  button {
    width: 100%;
  }
`;
