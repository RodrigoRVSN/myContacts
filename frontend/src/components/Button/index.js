import PropTypes from 'prop-types';
import { LoadingSpinner } from '../LoadingSpinner';
import { StyledButton } from './styles';

export default function Button({
  type, disabled, isLoading, children,
}) {
  return (
    <StyledButton
      disabled={disabled || isLoading}
      type={type}
    >
      {isLoading ? <LoadingSpinner /> : children}
    </StyledButton>
  );
}

Button.propTypes = {
  type: PropTypes.string,
  disabled: PropTypes.bool,
  isLoading: PropTypes.bool,
  children: PropTypes.node.isRequired,
};

Button.defaultProps = {
  type: 'button',
  disabled: false,
  isLoading: false,
};
