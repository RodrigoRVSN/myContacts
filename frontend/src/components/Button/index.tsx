import PropTypes from 'prop-types';
import { LoadingSpinner } from '../LoadingSpinner';
import { StyledButton } from './styles';

export default function Button({
  type, disabled, isLoading, children, onClick, danger,
}) {
  return (
    <StyledButton
      disabled={disabled || isLoading}
      type={type}
      onClick={onClick}
      danger={danger}
    >
      {isLoading ? <LoadingSpinner /> : children}
    </StyledButton>
  );
}

Button.propTypes = {
  type: PropTypes.string,
  disabled: PropTypes.bool,
  isLoading: PropTypes.bool,
  danger: PropTypes.bool,
  onClick: PropTypes.func,
  children: PropTypes.node.isRequired,
};

Button.defaultProps = {
  type: 'button',
  disabled: false,
  isLoading: false,
  danger: false,
  onClick: null,
};
