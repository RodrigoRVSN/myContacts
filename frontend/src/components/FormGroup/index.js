import React from 'react';
import PropTypes from 'prop-types';
import { Container } from './styles';

export function FormGroup({ children, error }) {
  return (
    <>
      <Container>
        {children}
        {error && <small>{error}</small>}
      </Container>
    </>
  );
}

FormGroup.propTypes = {
  children: PropTypes.node.isRequired,
  error: PropTypes.node,
};

FormGroup.defaultProps = {
  error: null,
};
