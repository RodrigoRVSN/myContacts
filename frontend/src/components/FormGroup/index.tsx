import { ReactNode } from 'react';
import { Container } from './styles';
import { LoadingSpinner } from '../LoadingSpinner';

type FormGroupProps = {
  children: ReactNode,
  error?: ReactNode,
  isLoading?: boolean,
}

export function FormGroup({ children, error = null, isLoading = false }: FormGroupProps) {
  return (
    <Container>
      <div className="form-item">
        {children}

        {isLoading && (
          <div className="loader-container">
            <LoadingSpinner />
          </div>
        )}
      </div>

      {error && <small>{error}</small>}
    </Container>
  );
}
