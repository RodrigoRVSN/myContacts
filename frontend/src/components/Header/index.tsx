import { Container } from './styles';

import logo from '../../assets/images/icons/logo.svg';

export default function Header() {
  return (
    <Container>
      <img src={String(logo)} alt="Logo MyContacts" />
    </Container>
  );
}
