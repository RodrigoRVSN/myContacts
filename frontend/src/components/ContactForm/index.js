import React, { useState } from 'react';
import PropTypes from 'prop-types';

import Button from '../Button';
import { FormGroup } from '../FormGroup';
import Input from '../Input';
import Select from '../Select';
import { ButtonContainer, Form } from './styles';
import isEmailValid from '../../utils/isEmailValid';
import useErrors from '../../hooks/useErrors';
import formatPhone from '../../utils/formatPhone';

export function ContactForm({ buttonLabel = '' }) {
  const [name, setName] = useState('');
  const [email, setEmail] = useState('');
  const [phone, setPhone] = useState('');
  const [category, setCategory] = useState('');

  const {
    setError,
    removeError,
    getErrorMessageByFieldName,
    errors,
  } = useErrors();

  const isFormValid = (name && errors.length === 0);

  function handleSubmit() {
    console.log('teste');
  }

  function handleNameChange(event) {
    setName(event.target.value);
    if (!event.target.value) {
      setError({ field: 'name', message: 'Nome é obrigatório' });
    } else {
      removeError('name');
    }
  }

  function handleEmailChange(event) {
    setEmail(event.target.value);
    if (event.target.value && !isEmailValid(event.target.value)) {
      setError({ field: 'email', message: 'E-mail inválido' });
    } else {
      removeError('email');
    }
  }

  function handlePhoneChange(event) {
    // phone.replace(/\D/g, '')
    setPhone(formatPhone(event.target.value));
  }

  function handleCategoryChange(event) {
    setCategory(event.target.value);
    if (event.target.value && !isEmailValid(event.target.value)) {
      setError({ field: 'email', message: 'E-mail inválido' });
    } else {
      removeError('email');
    }
  }

  return (
    <>
      <Form onSubmit={handleSubmit} noValidate>
        <FormGroup error={getErrorMessageByFieldName('name')}>
          <Input
            error={getErrorMessageByFieldName('name')}
            placeholder="Nome *"
            onChange={handleNameChange}
            value={name}
          />
        </FormGroup>

        <FormGroup error={getErrorMessageByFieldName('email')}>
          <Input
            type="email"
            error={getErrorMessageByFieldName('email')}
            placeholder="E-mail"
            onChange={handleEmailChange}
            value={email}
          />
        </FormGroup>

        <FormGroup>
          <Input
            type="text"
            placeholder="Telefone"
            onChange={handlePhoneChange}
            value={phone}
            maxLength="15"
          />
        </FormGroup>

        <FormGroup>
          <Select
            onChange={handleCategoryChange}
            value={category}
          >
            <option value="">Categoria</option>
            <option value="discord">Discord</option>
            <option value="instagram">Instagram</option>
          </Select>
        </FormGroup>

        <ButtonContainer>
          <Button type="submit" disabled={!isFormValid}>{buttonLabel}</Button>
        </ButtonContainer>
      </Form>
    </>
  );
}

ContactForm.propTypes = {
  buttonLabel: PropTypes.string,
};

ContactForm.defaultProps = {
  buttonLabel: '',
};
