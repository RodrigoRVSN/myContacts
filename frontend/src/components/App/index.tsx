// @ts-nocheck
import { BrowserRouter } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';
import GlobalStyles from '../../assets/styles/global';

import defaultTheme from '../../assets/styles/themes/default';
import RouteR from '../../Router';
import Header from '../Header';
import { ToastContainer } from '../Toast/ToastContainer';
import { Container } from './styles';

function App() {
  return (
    <BrowserRouter>
      <ThemeProvider theme={defaultTheme}>
        <GlobalStyles />

        <ToastContainer />

        <Container>
          <Header />
          <RouteR />
        </Container>
      </ThemeProvider>
    </BrowserRouter>
  );
}

export default App;
