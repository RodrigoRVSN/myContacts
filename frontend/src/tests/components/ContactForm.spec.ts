import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { ContactForm } from '../../components/ContactForm';

describe('<ContactForm />', () => {
  it('should render edit contacts page', () => {
    const { container } = render(<ContactForm />);
    expect(container).toMatchSnapshot();
  });

  it('should match the value typed and the expected', () => {
    const fn = jest.fn();
    render(<ContactForm handleNameChange={fn} />);

    expect(screen.queryByPlaceholderText(/Nome */i)).toBeInTheDocument();
    expect(screen.queryByPlaceholderText(/E-mail/i)).toBeInTheDocument();
    expect(screen.queryByPlaceholderText(/Telefone/i)).toBeInTheDocument();

    const inputName = screen.getByPlaceholderText(/Nome */i);
    const nameValue = 'nome digitado';
    userEvent.type(inputName, nameValue);
    expect(inputName.value).toBe(nameValue);

    const inputEmail = screen.getByPlaceholderText(/E-mail/i);
    const emailValue = 'rodrigo@email.com';
    userEvent.type(inputEmail, emailValue);
    expect(inputEmail.value).toBe(emailValue);

    const inputTel = screen.getByPlaceholderText(/Telefone/i);
    const telValue = '(15) 94002-8922';
    userEvent.type(inputTel, telValue);
    expect(inputTel.value).toBe(telValue);
  });

  it('should work correctly the button logic to disable', () => {
    const { getByRole } = render(<ContactForm />);

    const inputName = screen.getByPlaceholderText(/Nome */i);
    const nameValue = 'nome digitado';
    userEvent.type(inputName, nameValue);

    if (inputName.value !== '') {
      expect(getByRole('button')).not.toBeDisabled();
    } else {
      expect(getByRole('button')).toBeDisabled();
    }
  });
});
