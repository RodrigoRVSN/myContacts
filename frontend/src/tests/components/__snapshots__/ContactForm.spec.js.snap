// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContactForm /> should render edit contacts page 1`] = `
<div>
  <form
    class="sc-eCApnc hurtkR"
    novalidate=""
  >
    <div
      class="sc-gtsrHT jaRrUn"
    >
      <input
        class="sc-dlnjwi eBKfJg"
        placeholder="Nome *"
        value=""
      />
    </div>
    <div
      class="sc-gtsrHT jaRrUn"
    >
      <input
        class="sc-dlnjwi eBKfJg"
        placeholder="E-mail"
        type="email"
        value=""
      />
    </div>
    <div
      class="sc-gtsrHT jaRrUn"
    >
      <input
        class="sc-dlnjwi eBKfJg"
        maxlength="15"
        placeholder="Telefone"
        type="text"
        value=""
      />
    </div>
    <div
      class="sc-gtsrHT jaRrUn"
    >
      <select
        class="sc-hKFxyN bgquXc"
      >
        <option
          value=""
        >
          Categoria
        </option>
        <option
          value="discord"
        >
          Discord
        </option>
        <option
          value="instagram"
        >
          Instagram
        </option>
      </select>
    </div>
    <div
      class="sc-jSFjdj eIgFkL"
    >
      <button
        class="sc-bdnxRM ehnIfz"
        disabled=""
        type="submit"
      />
    </div>
  </form>
</div>
`;
